### PR TITLE
Automate binary release through github action

### LINKED ISSUE #22: Automate binary release through github action

### PR DESCRIPTION
**Pull request by messense**
This PR adds automated binary release to GitHub Release and PyPI.

Example run: https://github.com/messense/roapi/actions/runs/650339382

Fixes #22

### CODE REVIEW

**Comment by messense** [comment]
You can download and test the wheels: https://github.com/messense/roapi/suites/2252536813/artifacts/46855395

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+      - name: Upload wheels
+        uses: actions/upload-artifact@v2
+        with:
+          name: wheels
+          path: dist
+      - name: Archive binary
+        run: |
+          cd target/release
+          tar czvf roapi-http.tar.gz roapi-http
+          cd -
```
Comment (houqp) [inline .github/workflows/release.yml:49]: I think you can probably simplify these tar commands to:

```suggestion
        run:  tar czvf target/release/roapi-http.tar.gz -C target/release roapi-http
```

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,151 @@
+name: release
+
+on:
+  push:
+    branches: [ main ]
+    tags:
+  # pull_request:
+  #   branches: [ main ]
```
Comment (houqp) [inline .github/workflows/release.yml:8]: i think we should either remove these two lines or enable them

**Review by houqp — COMMENTED** [review]
Thanks @messense , love the idea of using pypi to help distribute the binaries :)

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+      - uses: actions/setup-python@v2
+        with:
+          python-version: 3.9
+      - name: Publish to PyPi
+        env:
+          TWINE_USERNAME: __token__
+          TWINE_PASSWORD: ${{ secrets.PYPI_PASSWORD }}
+        run: |
+          pip install --upgrade twine
+          twine upload --skip-existing *
```
[the remark this answers is not in the document]
Comment (messense) [reply .github/workflows/release.yml:151]: The current approach has the advantage of that you can download wheels for testing for any commit on master.

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+          pip install roapi-http --no-index --find-links dist --force-reinstall
+      - name: Build wheels - universal2
+        env:
+          DEVELOPER_DIR: /Applications/Xcode.app/Contents/Developer
+          MACOSX_DEPLOYMENT_TARGET: '10.9'
+        run: |
+          # set SDKROOT for C dependencies
+          export SDKROOT=$(xcrun --sdk macosx --show-sdk-path)
+          maturin build -m roapi-http/Cargo.toml -b bin --release --universal2 --out dist --no-sdist
+          pip install roapi-http --no-index --find-links dist --force-reinstall
```
[the remark this answers is not in the document]
Comment (messense) [reply .github/workflows/release.yml:39]: We are cross compiling on Linux jobs, only x86_64 can be installed and run so I simply omitted them altogether.

File: .github/workflows/release.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (houqp) [reply .github/workflows/release.yml:148]: i see, let's keep it then :)

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    steps:
+      - uses: actions/download-artifact@v2
+        with:
+          name: wheels
+      - uses: actions/setup-python@v2
+        with:
+          python-version: 3.9
+      - name: Publish to PyPi
+        env:
+          TWINE_USERNAME: __token__
```
Comment (messense) [inline .github/workflows/release.yml:139]: Note that you have to upload first to be able to generate a [api token](https://pypi.org/help/#apitoken)

Comment (messense) [reply .github/workflows/release.yml:139]: It can be easily done by downloading a wheel from CI artifacts and upload that.

**Review by houqp — APPROVED** [review]
LGTM

**Comment by houqp** [comment]
Thank you @messense!

**Comment by houqp** [comment]
As a follow up, it would be good to build with nightly toolchain and enable simd features for x86 platforms, which should give us a decent amount of performance boost.

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -0,0 +1,143 @@
+name: release
+
+on:
+  push:
+    branches: [ main ]
+    tags:
```
Comment (messense) [inline .github/workflows/release.yml:6]: The tag CI isn't trigger, maybe we need to write it as 

```yaml
on:
   push:
      branches: [ main ]
      tags: '*'
```

Comment (messense) [reply .github/workflows/release.yml:6]: https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#filter-pattern-cheat-sheet

Comment (houqp) [reply .github/workflows/release.yml:6]: Yes, can write it as `tags: [ 'roapi-http-v*' ]`? I am planning to release a columnq binary from this repo in the future for pure CLI usage as well. And possibly `roapi-grpc`. So would be good to namespace tags by sub project names.

something like this: https://github.com/delta-io/delta-rs/blob/main/.github/workflows/python_release.yml#L5

**Pull request merged**